feat: add schema migrations framework

bloodearnest opened the pull request:
We have some pending db changes to make. In preparation for this, this
change adds a formal migration system to the micro-ORM.

 * To avoid having to package raw sql files, we track the migrations as
   string literals in python code, as dict mapping version to SQL
   string.

 * We use the `PRAGMA user_version` db variable to track the migration
   version, using transactions to ensure it is set appropriately.

 * We add a migrate command, analagous to django `manage.py migrate`.
   This will ensure a db exists and is at the latest schema version.
   This is the only way to create/migrate a db, and should be run as an
   explicit step as part of production deployments

 * Previously, in production schema creation happened lazily when the
   service started up. Extending this to apply migrations automatically
   on service start up seemed risky. Instead we add an explict check
   that the db exists and is up to date, or else we do not start up.
   This is designed to catch configuration issues or unapplied
   migrations in production.

 * For `jobrunner.cli.local_run` however, we do automatically apply the
   migrations on behalf of the user, as each project has its own db. The
   data they have is an optimisation, rather than a requirement, as they
   can always re-run the entire project if needed.

 * Note: there are no actual migrations in this initial change - the
   migration framework will be used in subsequent PRs

ghickman commented on DEVELOPERS.md:
````diff
@@ -163,3 +163,47 @@ you can do so:
+""")
+```
+
+These statements are run together in a single transaction, allong with
````
```suggestion
These statements are run together in a single transaction, along with
```

bloodearnest replied:
fixed

ghickman commented on jobrunner/lib/database.py:
```diff
@@ -31,6 +32,12 @@ def databaseclass(cls):
 
+def migration(version, sql):
+    """Used to record a migration"""
+    assert version not in MIGRATIONS, f"Migration {version} alread exists."
```
```suggestion
    assert version not in MIGRATIONS, f"Migration {version} already exists."
```

bloodearnest replied:
fixed

ghickman commented on jobrunner/lib/database.py:
```diff
@@ -153,37 +164,104 @@ def get_connection():
-    schema_count = list(conn.execute("SELECT COUNT(*) FROM sqlite_master"))[0][0]
-    if schema_count == 0:
+
+    latest_version = max(migrations) if migrations else 0
```
You could also use the default kwarg here:

    `max(migrations, default=0)`

if you wanted to drop the conditional.  I'm not sure it adds much value though.

bloodearnest replied:
Thats more readable, thanks

ghickman commented on pyproject.toml:
```diff
@@ -25,6 +25,10 @@ add_job = "jobrunner.cli.add_job:run"
 kill_job = "jobrunner.cli.kill_job:run"
 local_run = "jobrunner.cli.local_run:run"
 retry_job = "jobrunner.cli.retry_job:run"
+prepare_for_reboot = "jobrunner.cli.prepare_for_reboot:run"
+flags = "jobrunner.cli.flags:run"
+migrate = "jobrunner.cli.migrate:run"
```
Could you alphabetise these? Makes them a little easier to skim 🙂

bloodearnest replied:
done

ghickman reviewed (approved):
Looks great 👍

Couple of suggestions you might find useful.

evansd commented on jobrunner/service.py:
```diff
@@ -32,7 +32,7 @@ def main():
     # ensure db is set up before we start any threads, otherwise if we don't already have
     # a database (which we don't in tests) then threads will race to create the schema
-    get_connection()
+    init_db()
```
This call means that if the database file isn't in the place we expect it we'll just create a new empty file and carry on without error. But I thought the idea was that we'd fail hard unless the database file had already been created?

bloodearnest replied:
Hmm, yes we did discuss that. But my main concern was to detect when migrations needed applying, so that got dropped. And ever

It's a bit messy, as it means init_db will need two modes when a db doesn't exist one where it fails (which is only used here) and one where it doesn't (which is used everywhere else, i.e. tests and local_run.

And how do we know if this is a new backend or not? We'd need a way to create the initial db so this check wouldn't error.  The current migrate command doesn't do that, it assumes a db exists. 

So, if we use django as our model, then an initial migrate *is* needed before a service will run successfully in a new environment. So perhaps the migrate command should do that.

We'll still need to two modes though, or possibly two different functions.

bloodearnest replied:
Have updated to implement the above

evansd commented on jobrunner/lib/database.py:
```diff
@@ -153,37 +164,104 @@ def get_connection():
+    pass
+
+
+def init_db(filename=None, migrations=None):
```
Related to the comment below, it looks like `init_db()` bundles up both opening the connection and creating the database file if it doesn't exist. But I thought we wanted to separate those?

bloodearnest commented:
Updated PR to answer @dave's point and to clarify the "manual" step.

Basically, to create a db, we now have to run migrate first, just like django.

Added a start up check that the db is in the right state, to be cautious initially.

Also implemented a slightly nicer in memory db set up.
